File: solutions/JavaScript/generate_parentheses_backtracking.js
function generateParenthesis(n) {
    const result = [];

    function backtrack(current, left, right) {
        if (current.length === 2 * n) {
            result.push(current);
            return;
        }
        if (left < n) backtrack(current + "(", left + 1, right);
        if (right < left) backtrack(current + ")", left, right + 1);
    }

    backtrack("", 0, 0);
    return result;
}

File: solutions/Rust/generate_parentheses_backtracking.rs
pub fn generate_parenthesis(n: i32) -> Vec<String> {
    let mut result = Vec::new();
    fn backtrack(result: &mut Vec<String>, current: String, left: i32, right: i32, n: i32) {
        if current.len() == (2 * n) as usize {
            result.push(current);
            return;
        }
        if left < n {
            backtrack(result, current.clone() + "(", left + 1, right, n);
        }
        if right < left {
            backtrack(result, current + ")", left, right + 1, n);
        }
    }
    backtrack(&mut result, String::new(), 0, 0, n);
    result
}
